doc/roadmap-2024.md

```diff
@@ -0,0 +1,75 @@
+
+### Low Read Latency of Hot Region
+
+Read queries being evenly distributed in different data ranges is an ideal workload, which can make good use of all resources across different TiKV nodes. But in the real world, queries tend to skew to some small data ranges, like latest added/modified data has higher possibility to be retrieved, a specified category's data has higher read queries in some special time range, etc. We call these highly retrieved data ranges as hot regions. Hot regions tend to cause higher query latency since there might be some queries need to wait. A faster in-memory engine cache layer might have a great benefit in this hot region case since it will use less resource to serve a query.
+
+### Data Warmup before Leader Transfer
+
+When the leader of a data range transferred to another TiKV node, because the data of this range is not in memory yet in the new TiKV node, if there are some read queries route to this new leader, it tends to cause higher read latency for these queries. Scenarios like leader re-balancing, TiKV nodes rolling restart may cause leader transfer. We probably need a data warm-up mechanism for the new leader before transfer the leader role to it to reduce the read latency in such scenario.
+
+### Stable Latency During Planned Operations
+
+As a distributed system, there are some planned operations may happen, like rolling restart, rolling upgrade, scale-in some TiKV nodes, scale-out some TiKV nodes, etc. As the cornerstone of other mission critical databases like TiDB, TiKV should provide continuous service and stable latency even when there is a planned operation.
+
+### Stable Latency Under High Data Density Scenario
+
+Some customers' data volume of single TiKV cluster reach as large as 500TB, and each TiKV node carries more than 4TB data. These large clusters experienced some latency jitter issues, those issues caused by reasons like too many regions in a single TiKV node, background jobs cause resources competition, large compaction jobs, impact of large RocksDB instance's global mutex, etc. We are going to eliminate or mitigate these latency jitter issues, and provide stable latency in such high data density scenario.
+
+## Comprehensive Memory Management
+
+### Component-level Memory Trace
+
+Introduce better memory trace for different internal components like transaction module, read pool, raftstore, etc. In this way, we can monitor the memory usage of different components.
+
+### Memory Quota Limit
+
+TiKV has some soft limits for different components' memory usage like scheduler, raftstore, but there are still some components like coprocessor don't have memory quota limit. In the coprocessor, we use the amount of pending queries to throttle the resources usage of coprocessor, but this still can result in Out-Of-Memory(OOM) when there are too many "BIG" queries which retrieve a large amount of data.
+
+## Powerful and Robust Storage Engine
```

Comment:
May be add experimenting with using io_uring based https://github.com/bytedance/monoio to replace sync IO for raft engine.

Reply:
The trick part is that we already separate the writing raft-engine operations asynchronously from the main message handling procedure, how much extra benefit to make the write raft-engine asynchronous one more step?

Reply:
cc @siddontang who suggested this. the reason is that writing to file is not fully asynchronous unless one uses uring. monoio is more or less a drop in replacement, so it should be possible to investigate this.

Reply:
@afeinberg io_uring is just a research, depending on our bandwidth, we don't need to write it down in the roadmap.